objects/src/accounts/vault.rs

```diff
@@ -59,11 +65,161 @@ impl AccountVault {
+
+        // if the asset already exists, return an error
+        if old != [ZERO, ZERO, ZERO, ZERO] {
+            return Err(AccountError::DuplicateNonFungibleAsset(asset));
+        }
+
+        Ok(asset)
+    }
+
+    // REMOVE ASSET
+    // --------------------------------------------------------------------------------------------
+    /// Remove the specified asset from the vault.
+    ///
+    /// # Errors
+    /// - The fungible asset is not found in the vault.
+    /// - The amount of the fungible asset in the vault is less than the amount to be removed.
+    /// - The non-fungible asset is not found in the vault.
+    pub fn remove_asset(&mut self, asset: Asset) -> Result<Asset, AccountError> {
+        Ok(match asset {
+            Asset::Fungible(asset) => Asset::Fungible(self.remove_fungible_asset(asset)?),
+            Asset::NonFungible(asset) => Asset::NonFungible(self.remove_non_fungible_asset(asset)?),
+        })
+    }
+
+    /// Remove the specified fungible asset from the vault.
+    ///
+    /// # Errors
+    /// - The asset is not found in the vault.
+    /// - The amount of the asset in the vault is less than the amount to be removed.
+    fn remove_fungible_asset(
+        &mut self,
+        asset: FungibleAsset,
+    ) -> Result<FungibleAsset, AccountError> {
+        // fetch the asset from the vault.
+        let mut current: FungibleAsset = match self.asset_tree.get_value(asset.key().into()) {
+            current if current == [ZERO, ZERO, ZERO, ZERO] => {
+                return Err(AccountError::FungibleAssetNotFound(asset))
+            }
+            current => current.try_into().expect("tree only contains valid assets"),
+        };
+
+        // subtract the amount of the asset to be removed from the current amount.
+        current
+            .sub(asset.amount())
+            .map_err(AccountError::SubtractFungibleAssetBalanceError)?;
+
+        // if the amount of the asset is zero, remove the asset from the vault.
+        let new = match current.amount() {
+            0 => {
+                // return ZERO value to insert into the vault
+                [ZERO, ZERO, ZERO, ZERO]
+            }
+            _ => current.into(),
+        };
+        self.asset_tree.insert(asset.key().into(), new);
+
+        // return the asset that was removed.
+        Ok(asset)
+    }
+
+    /// Remove the specified non-fungible asset from the vault.
+    ///
+    /// # Errors
+    /// - The non-fungible asset is not found in the vault.
+    fn remove_non_fungible_asset(
+        &mut self,
+        asset: NonFungibleAsset,
+    ) -> Result<NonFungibleAsset, AccountError> {
+        // remove the asset from the vault.
+        let old = self.asset_tree.insert(asset.key().into(), [ZERO, ZERO, ZERO, ZERO]);
+
+        // return an error if the asset did not exist in the vault.
+        if old == [ZERO, ZERO, ZERO, ZERO] {
+            return Err(AccountError::NonFungibleAssetNotFound(asset));
+        }
```

Comment:
Similar comment to the one before: `TieredSmt` doesn't handle deletions correctly - so, this may result in incorrect results.

Reply:
Will add a TODO.